### PR TITLE
JVMChaos: minor refine with byteman-helper

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
Signed-off-by: xiang <xiang13225080@163.com>

### What problem does this PR solve?

already implement the stress and gc with [byteman-helper](https://github.com/chaos-mesh/byteman-helper), this pr refine JVMChaos with it.
1. need to submit `byteman-helper.jar`
2. refine some code by the way

### Related changes

- [ ] Need to update `chaos-mesh/website`
- [ ] Need to update `Dashboard UI`
- Need to **cheery-pick to release branches**
  - [x] release-2.1
  - [ ] release-2.0

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [ ] No code
- [ ] Manual test (add steps below)

Side effects

- [ ] Breaking backward compatibility

### Release note 

```text
Please add a release note.

You can safely ignore this section if you don't think this PR needs a release note.
```

### DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```

### CODE REVIEW

File: pkg/chaosdaemon/jvm_server.go [hunk 1]
```diff
@@ -123,13 +121,28 @@ func (s *DaemonServer) InstallJVMRules(ctx context.Context,
...
+
 	// submit rules
 	filename, err := writeDataIntoFile(req.Rule, "rule.btm")
 	if err != nil {
 		return nil, err
 	}
 
-	bmSubmitCmd := fmt.Sprintf(bmSubmitCommand, req.Port, "l", filename)
+	bmSubmitCmd = fmt.Sprintf(bmSubmitCommand, req.Port, "l", filename)
 	processBuilder = bpm.DefaultProcessBuilder("sh", "-c", bmSubmitCmd).SetContext(ctx)
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/chaosdaemon/jvm_server.go:146]: ignore

File: pkg/time/time_linux_arm64.go [hunk 2]
```diff
@@ -16,13 +16,14 @@
...
 import (
+	"github.com/go-logr/logr"
 	"github.com/pkg/errors"
 
 	"github.com/chaos-mesh/chaos-mesh/pkg/mock"
 )
 
 // ModifyTime modifies time of target process
-func ModifyTime(pid int, deltaSec int64, deltaNsec int64, clockIdsMask uint64) error {
+func ModifyTime(pid int, deltaSec int64, deltaNsec int64, clockIdsMask uint64, logger logr.Logger) error {
```
Comment (Andrewmatilde) [inline pkg/time/time_linux_arm64.go:26]: PLZ , don't do this here, there are about 2 PR doing the same thing.

File: api/v1alpha1/jvmchaos_types.go [hunk 3]
```diff
@@ -88,19 +86,39 @@ type JVMParameter struct {
 	LatencyDuration int `json:"latency"`
 
 	// +optional
-	// the CPU core number need to use, only set it when action is stress
-	CPUCount int `json:"cpuCount"`
+	RuleData string `json:"ruleData"`
```
Comment (Andrewmatilde) [inline api/v1alpha1/jvmchaos_types.go:89]: PLZ add comment here.

**Review by Andrewmatilde — APPROVED** [review]
LGTM

File: api/v1alpha1/jvmchaos_types.go [hunk 4]
```diff
@@ -63,17 +63,15 @@ const (
...
 type JVMParameter struct {
-	// +optional
-	// byteman rule name, should be unique, and will use JVMChaos' name if not set
-	Name string `json:"name"`
+	JVMCommonSpec `json:",inline"`
 
-	// +optional
-	// Java class
-	Class string `json:"class"`
+	JVMClassMethodSpec `json:",inline"`
```
Comment (cwen0) [inline api/v1alpha1/jvmchaos_types.go:68]: Will these changes break compatibility?

Comment (cwen0) [reply api/v1alpha1/jvmchaos_types.go:68]: Oh, my fault, these changes won't break compatibility.

File: api/v1alpha1/jvmchaos_types.go (hunk 3 above)
Comment (cwen0) [reply api/v1alpha1/jvmchaos_types.go:89]: @WangXiangUSTC PTAL

File: api/v1alpha1/jvmchaos_types.go [hunk 5]
```diff
@@ -88,19 +86,39 @@ type JVMParameter struct {
...
 
+// JVMCommonSpec is the common specification for JVMChaos
+type JVMCommonSpec struct {
 	// +optional
-	// the memory type need to locate, only set it when action is stress, the value can be 'stack' or 'heap'
-	MemoryType string `json:"memType"`
+	// the port of agent server, default 9277
+	Port int32 `json:"port,omitempty"`
 
+	// the pid of Java process which need to attach
```
Comment (fewdan) [inline api/v1alpha1/jvmchaos_types.go:98]: Better to change it to "which needs to xxx"

File: api/v1alpha1/jvmchaos_types.go [hunk 6]
```diff
@@ -88,19 +86,39 @@ type JVMParameter struct {
...
 	// +optional
-	RuleData string `json:"ruleData"`
+	// the method in Java class
+	Method string `json:"method,omitempty"`
+}
+
+// JVMStressSpec is the specification for stress
+type JVMStressCfgSpec struct {
+	// +optional
+	// the CPU core number need to use, only set it when action is stress
```
Comment (fewdan) [inline api/v1alpha1/jvmchaos_types.go:116]: "needs" will be better.

File: api/v1alpha1/jvmchaos_types.go [hunk 7]
```diff
@@ -88,19 +86,39 @@ type JVMParameter struct {
...
+	CPUCount int `json:"cpuCount,omitempty"`
+
+	// +optional
+	// the memory type need to locate, only set it when action is stress, the value can be 'stack' or 'heap'
```
Comment (fewdan) [inline api/v1alpha1/jvmchaos_types.go:120]: ditto

File: api/v1alpha1/physical_machine_chaos_types.go [hunk 8]
```diff
@@ -387,6 +371,16 @@ type JVMExceptionSpec struct {
 	ThrowException string `json:"exception,omitempty"`
 }
 
+type JVMStressSpec struct {
+	JVMCommonSpec `json:",inline"`
+
+	// the CPU core number need to use, only set it when action is stress
```
Comment (fewdan) [inline api/v1alpha1/physical_machine_chaos_types.go:377]: ditto

File: api/v1alpha1/physical_machine_chaos_types.go [hunk 9]
```diff
@@ -387,6 +371,16 @@ type JVMExceptionSpec struct {
...
+	CPUCount int `json:"cpu-count,omitempty"`
+
+	// the memory type need to locate, only set it when action is stress, the value can be 'stack' or 'heap'
```
Comment (fewdan) [inline api/v1alpha1/physical_machine_chaos_types.go:380]: ditto

File: controllers/chaosimpl/jvmchaos/impl.go [hunk 10]
```diff
@@ -161,46 +190,57 @@ func generateRuleData(spec *v1alpha1.JVMChaosSpec) error {
...
+		// the bind and condition is useless, only used for fill the template
+		bytemanTemplateSpec.Bind = "flag:boolean=true"
+		bytemanTemplateSpec.Condition = "true"
+		if spec.CPUCount > 0 {
+			bytemanTemplateSpec.Do = fmt.Sprintf("injectCPUStress(\"%s\", %d)", spec.Name, spec.CPUCount)
 		} else {
-			ruleParameter.StressType = "MEMORY"
-			ruleParameter.StressValueName = "MEMORYTYPE"
-			ruleParameter.StressValue = ruleParameter.MemoryType
+			bytemanTemplateSpec.Do = fmt.Sprintf("injectMemStress(\"%s\", %s)", spec.Name, spec.MemoryType)
```
Comment (fewdan) [inline controllers/chaosimpl/jvmchaos/impl.go:216]: If CPUCount>0 and MemoryType is not empty, this will only inject MemStress, is this expected?

**Review by fewdan — APPROVED** [review]
rest LGTM

File: controllers/chaosimpl/jvmchaos/impl.go (hunk 10 above)
Comment (fewdan) [reply controllers/chaosimpl/jvmchaos/impl.go:216]: I took a look and the existing logic is like this. Is it necessary to support injection of CPUStress and MEMStress at the same time in the future?

Comment (WangXiangUSTC) [reply controllers/chaosimpl/jvmchaos/impl.go:216]: There is a validation in the webhook, can only set one of `CPUCount` and `MemoryType`. In my opinion, it's not necessary to inject both of them in one experiment.
https://github.com/chaos-mesh/chaos-mesh/blob/b4b33740bb6dbcf93892d693adc32376f9bd5a18/api/v1alpha1/jvmchaos_webhook.go#L51

**Comment by WangXiangUSTC** [comment]
@fewdan add comments addressed, PTAL again

**Review by fewdan — APPROVED** [review]
LGTM, thanks.

**Pull request merged**